# FREEC-Plot - Fork
Plotting outputs of Control-FREEC for targeting sequencing data in specific chromosome or gene

    Example: Rscript plot_chr8.R -i inputfolder -o outputfolder -b bedfile -r -f
    Usage: plot_chr8.R [-[-help|h]] [-[-input|i] <character>] [-[-ratioGraph|r]] [-[-BAFGraph|f]] [-[-output|o] <character>] [-[-gene|g] <character>] [-[-chr|c] <character>] [-[-bed|b] <character>]
    -h|--help          help manual
    -i|--input         filepath for *.BAF.txt
    -r|--ratioGraph    if RatioGraph should be drawn
    -f|--BAFGraph      if BAFGraph should be drawn
    -o|--output        filepath for output figures
    -g|--gene          gene/transcript symbol for output figures
    -c|--chr           chrmosome for output figures
    -y|--cytoBand      UCSC cytoBand for output figures
    -b|--bed           panel bed file

# Instructions
* For ratiaGraph, ratio.txt & CNV files generated by Control-FREEC have to be contained in the inputfolder.
* For BAFGraph, BAF.txt need to be provided. 
* The x-axis of the plot represents bases covered by the targeted regions.
* mark CopyNumber > 2 log2 ratio in firebrick3 color
* mark CopyNumber < 2 log2 ratio in mediumblue color
* mark CopyNumber = 2 log2 ratio in green

# Merge output Excels using merge_excel_by_sheet_name.R
* multi-sample outputs could be merged into one table
